blog/02-15-2023-build-feature-rich-persistent-dapp-on-celo-using-wagmi/build-feature-rich-persistent-dapp-on-celo-using-wagmi.md

```diff
@@ -0,0 +1,747 @@
+
+## Prerequisites​
+
+We will walk through one of the many ways of connecting and interacting with dapp on the Celo blockchain using wagmi. Celo blockchain is Ethereum compatible, and this indicates that we can use the same tools for building dapps on the Ethereum blockchain to build for Celo. This is an advanced tutorial, hence it is expected that you know the following:
+
+- Solidity development
+- Javascript and typescript
+- How to use hardhat for smart contract development. Refer to **[this](https://docs.celo.org/blog/tutorials/getting-started-on-celo-with-hardhat)** and **[this](https://docs.celo.org/blog/tutorials/advance-hardhat-configuration-on-celo-using-plugins)** to learn how to use hardhat.
+
+## Requirements​
+
+Install the following tools:
+
+- For an editor. I recommend VSCode.
+- NodeJs version >=14.0.0. I use version 18.12.1
+- [Setup and configure hardhat](https://docs.celo.org/blog/tutorials/advance-hardhat-configuration-on-celo-using-plugins).
+
+**What we are building**
+
+We will create a decentralized application for swapping ERC20 tokens for $Celo. For a full description, please refer to the **[readme](https://github.com/bobeu/feature-rich-persistent-dapp-on-celo-using-wagmi/README.md)**
+
+The dApp was built in two parts.
+
+1. Smart contracts
+2. Frontend
+
+**Smart contracts**
+
+- Clone project
+
+```bash
+git clone https://github.com/bobeu/feature-rich-persistent-dapp-on-celo-using-wagmi.git
+```
+- Navigate to the root folder
+
+```bash
+cd feature-rich-persistent-dapp-on-celo-using-wagmi/backend
+```
+
+- Install dependencies
+
+```bash
+yarn install
+```
+Now, let's examine the core functionalities.
+
+**Backend file/folder structure**
+
+- contracts
+- deploy
+  - 00_deploy.js
+- scripts
+  - deploy.ts
+- test
+  - swapLab-test.ts
+- _gitignore_
+- _hardhat.config.ts_
+- _package-lock.json_
+- _package.json_
+- _README.md_
+- _tsconfig.json_
+- _yarn.lock_
+
+**Contract folder**
+
+- `SwapLab.sol`
+  - For top-level ownership of the contract, we'll use the openzeppellin's _Ownable.sol_.
+
+  ```js
+  import "@openzeppelin/contracts/access/Ownable.sol";
+  ```
+
+  - _IERC20.sol_ to quickly create an ERC20 compatible token.
+  
+  ```js
+  import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+  ```
+
+  - _IERC20Metadata.sol_ is an interface for token metadata such as token names, symbols, and decimals.
+
+  ```js
+  import "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+  ```
+
+  - To safely perform simple arithmetic calculations, and avoid underflow or overflow risk in contracts, we import the _SafeMath.sol_. Although this may not be necessary for recent solidity versions since they come with automated built-in checks.
```

Comment:
- To safely perform simple arithmetic calculations and avoid underflow or overflow risk in contracts, we import the `_SafeMath.sol_`. However, this may not be necessary for recent solidity versions since they come with built-in automated checks.